### PR TITLE
Fixes #1254 - Add flag to define Marathon PORT prefix.

### LINKED ISSUE #1254: Add flag to define Marathon PORT prefix
Marathon overrrides the PORT env inside the container. This can be a nice feature but is also causing trouble if you need to connect to the port the container app did set to PORT.

A flag to set env prefix for the ports marathon sets would solve this.

### PR DESCRIPTION
**Pull request by gkleiman**

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala [hunk 1]
```diff
@@ -91,7 +91,7 @@ case class AppDefinition(
   }
 
   def toProto: Protos.ServiceDefinition = {
-    val commandInfo = TaskBuilder.commandInfo(this, None, None, Seq.empty)
+    val commandInfo = TaskBuilder.commandInfo(this, None, None, Seq.empty, None)
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: Named a parameters would be nice here. Intellij can do that automatically.

File: src/main/scala/mesosphere/mesos/TaskBuilder.scala [hunk 2]
```diff
@@ -178,12 +179,16 @@ class TaskBuilder(app: AppDefinition,
...
+                  taskId: Option[TaskID],
+                  host: Option[String],
+                  ports: Seq[Long],
+                  envPrefix: Option[String]): CommandInfo = {
     val containerPorts = for (pms <- app.portMappings) yield pms.map(_.containerPort)
     val declaredPorts = containerPorts.getOrElse(app.ports)
     val envMap: Map[String, String] =
       taskContextEnv(app, taskId) ++
-        portsEnv(declaredPorts, ports) ++ host.map("HOST" -> _) ++
+        addPrefix(envPrefix, portsEnv(declaredPorts, ports)) ++ host.map("HOST" -> _) ++
```
Comment (kolloch) [inline src/main/scala/mesosphere/mesos/TaskBuilder.scala]: Mention "HOST" as exception in the documentation?

**Comment by kolloch** [comment]
Explain/Reference new configuration option here as well: task-environment-vars.md

File: src/test/scala/mesosphere/mesos/TaskBuilderTest.scala [hunk 3]
```diff
@@ -572,7 +572,8 @@ class TaskBuilderTest extends MarathonSpec {
         ),
         Some(TaskID("task-123")),
         Some("host.mega.corp"),
-        Seq(1000, 1001)
+        Seq(1000, 1001),
+        None
```
Comment (kolloch) [inline src/test/scala/mesosphere/mesos/TaskBuilderTest.scala]: use named arguments at least for things like "None"

File: src/test/scala/mesosphere/mesos/TaskBuilderTest.scala [hunk 4]
```diff
@@ -633,6 +636,32 @@ class TaskBuilderTest extends MarathonSpec {
...
+    assert("1000" == env("CUSTOM_PREFIX_PORT"))
+
+    assert(!env.contains("PORT0"))
+    assert("1000" == env("CUSTOM_PREFIX_PORT0"))
+    assert("1000" == env("CUSTOM_PREFIX_PORT_8080"))
+
+    assert(!env.contains("PORT1"))
+    assert("1001" == env("CUSTOM_PREFIX_PORT1"))
+    assert("1001" == env("CUSTOM_PREFIX_PORT_8081"))
+  }
```
Comment (kolloch) [inline src/test/scala/mesosphere/mesos/TaskBuilderTest.scala]: Explicitly check for all Variables not having the prefix and compare it against a list of exceptions.

That's more future proof in my opinion.

**Comment by kolloch** [comment]
Also add a changelog entry.

File: src/main/scala/mesosphere/mesos/TaskBuilder.scala [hunk 5]
```diff
@@ -267,6 +272,13 @@ object TaskBuilder {
     }
   }
 
+  def addPrefix(envVarsPrefix: Option[String], env: Map[String, String]): Map[String, String] = {
```
Comment (kolloch) [inline src/main/scala/mesosphere/mesos/TaskBuilder.scala]: You COULD write this as:

```scala
envVarsPrefix.fold(env)(prefix => env.map { case (k,v) => (prefix + k) -> v })
```

but your version is fine, too

**Comment by gkleiman** [comment]
Here goes an updated version.

If you like this, then I'll write something for the changelog and ask Matthias to add it to mv/changelog-0.10.

File: src/test/scala/mesosphere/mesos/TaskBuilderTest.scala [hunk 6]
```diff
@@ -619,24 +621,55 @@ class TaskBuilderTest extends MarathonSpec {
...
+    assert("1000" == env("CUSTOM_PREFIX_PORT0"))
+    assert("1000" == env("CUSTOM_PREFIX_PORT_8080"))
+
+    assert("1001" == env("CUSTOM_PREFIX_PORT1"))
+    assert("1001" == env("CUSTOM_PREFIX_PORT_8081"))
+
+    assert("host.mega.corp" == env("CUSTOM_PREFIX_HOST"))
+
+    assert(Seq("HOST", "PORTS", "PORT0", "PORT1").forall(k => !env.contains(k)))
+    assert(Seq("MESOS_TASK_ID", "MARATHON_APP_ID", "MARATHON_APP_VERSION").forall(env.contains))
```
Comment (kolloch) [inline src/test/scala/mesosphere/mesos/TaskBuilderTest.scala]: I would rather check for all variables that do not start with the prefix and ascertain that all exceptions are in the list.

If you add an unprefixed variable now, no test will fail.

**Comment by kolloch** [comment]
LGTM. Don't forget the changelog!

**Pull request merged**